### PR TITLE
Added capability of regex matching audio device names https://github.com/GrandOrgue/grandorgue/issues/1265

### PR DESCRIPTION
**Pull request by oleg68**
Resolves: #1265.

This is the last PR related to #1265.

Now there is a `Matching` button on the Audio Settings tab. It allows to set up a regex pattern for handling bounced device names.

This PR also sets some default regex pattern for PortAudio/Alsa

### CODE REVIEW

File: help/grandorgue.xml [hunk 1]
```diff
@@ -4770,6 +4785,82 @@ role="strong">STOP</emphasis> button is depressed.
         </sect3>
       </sect2>
     </sect1>
+    <sect1 id="devicematching">
+        <title>Devices Matching</title>
```
Comment (larspalo) [inline help/grandorgue.xml:4789]:
```suggestion
        <title>Device Matching</title>
```

File: help/grandorgue.xml [hunk 2]
```diff
@@ -4770,6 +4785,82 @@ role="strong">STOP</emphasis> button is depressed.
...
+          This dialog allows to define matching rules for audio/midi device
+          names  
+        </para>
+        <variablelist>
+          <varlistentry>
+            <term>Physical device name</term>
+            <listitem>
+              <para>
+                This is the name of a Audio/MIDI device as it is reported by
+                the Sound/MIDI port. A user can not alter it, but this name
```
Comment (larspalo) [inline help/grandorgue.xml:4811]:
```suggestion
                the Sound/MIDI port. A user can not alter it, and this name
```

File: help/grandorgue.xml [hunk 3]
```diff
@@ -4770,6 +4785,82 @@ role="strong">STOP</emphasis> button is depressed.
...
+                <simpara>The logical device name must be unique</simpara>
+              </note>
+              <note>
+                <simpara>
+                  If the user alters the logical name, a regex pattern must
+                  be provided so that GrandOrgue can find this device by its
+                  physical name.
+                </simpara>
+                <simpara>
+                  Specify there the most common part among the physical names
```
Comment (larspalo) [inline help/grandorgue.xml:4834]:
```suggestion
                  Specify the most common part among the different physical names
```

File: help/grandorgue.xml [hunk 4]
```diff
@@ -4770,6 +4785,82 @@ role="strong">STOP</emphasis> button is depressed.
...
+                  of the device appear across reboots
```
Comment (larspalo) [inline help/grandorgue.xml:4835]:
```suggestion
                  that the device name can have across reboots
```

File: help/grandorgue.xml [hunk 5]
```diff
@@ -4770,6 +4785,82 @@ role="strong">STOP</emphasis> button is depressed.
...
+                the Sound/MIDI port. A user can not alter it, but this name
+                may be not persistent across reboots
+              </para>
+            </listitem>
+          </varlistentry>
+          <varlistentry>
+            <term>Logical device name</term>
+            <listitem>
+              <para>
+                This is the name of MIDI device as it is presented to the user.
```
Comment (larspalo) [inline help/grandorgue.xml:4820]:
```suggestion
                This is the name of Audio/MIDI device as it is presented to the user.
```

File: help/grandorgue.xml [hunk 6]
```diff
@@ -4770,6 +4785,82 @@ role="strong">STOP</emphasis> button is depressed.
...
+                  of the device appear across reboots
+                </simpara>
+              </note>
+            </listitem>
+          </varlistentry>
+          <varlistentry>
+            <term>Physical name regex pattern</term>
+            <listitem>
+              <para>
+                This is a pattern that allows GrandOrgue to find this MIDI
```
Comment (larspalo) [inline help/grandorgue.xml:4844]:
```suggestion
                This is a pattern that allows GrandOrgue to find this Audio/MIDI
```

File: help/grandorgue.xml [hunk 7]
```diff
@@ -4770,6 +4785,82 @@ role="strong">STOP</emphasis> button is depressed.
...
+                device with it's physical device name.
+              </para>
+              <para>
+                The simplest pattern would be the most common part of the
+                physical device name that is present always and it kept
```
Comment (larspalo) [inline help/grandorgue.xml:4849]:
```suggestion
                physical device name that is always present and kept
```

File: help/grandorgue.xml [hunk 8]
```diff
@@ -4770,6 +4785,82 @@ role="strong">STOP</emphasis> button is depressed.
...
+                across reboots. More advanced regular expressions are also
+                supported, look at
+                <ulink url="https://en.wikibooks.org/wiki/Regular_Expressions/POSIX_Basic_Regular_Expressions">
+                  the full syntax
```
Comment (larspalo) [inline help/grandorgue.xml:4853]:
```suggestion
                  for the full syntax.
```

File: src/grandorgue/sound/GOSoundDevInfo.cpp [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    apiName,
+    name,
+    GOSoundPortFactory::getFullDeviceName(portName, apiName, name),
+    channels,
+    isDefault,
+    fillLogicalNameAndRegEx) {}
+
+/* Don't use the short constructor here because it calls to
+ * GOSoundPortFactory::getFullDeviceName that uses static variables from other
+ * units. The orger of initializing static variables in differed units is not
```
Comment (larspalo) [inline src/grandorgue/sound/GOSoundDevInfo.cpp:50]:
```suggestion
 * units. The order of initializing static variables in differed units is not
```

**Comment by oleg68** [comment]
@larspalo I corrected this PR as you suggested.

**Review by larspalo — APPROVED** [review]
I've tested some 20+ reboots with this PR and it just works in every one of them!

**Review by rousseldenis — APPROVED** [review]

**Pull request merged**